modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/agg/AggregateRow.java

```diff
@@ -73,9 +102,29 @@ public Object[] createOutput(ImmutableBitSet allFields, byte groupId) {
     /** Writes aggregate state of the given row to given array. */
     public void writeTo(Object[] output, ImmutableBitSet allFields, byte groupId) {
         int cardinality = allFields.cardinality();
+
+        AccumulatorsState result = new AccumulatorsState(accs.size());
+
         for (int i = 0; i < accs.size(); i++) {
-            AccumulatorWrapper<RowT> wrapper = accs.get(i);
-            output[i + cardinality] = wrapper.end();
+            AccumulatorWrapper<RowT> acc = accs.get(i);
+
+            state.setIndex(i);
+            result.setIndex(i);
+
+            if (acc.isDistinct()) {
+                Set<Object> distinctSet = distinctSets.get(i);
+
+                for (var arg : distinctSet) {
+                    acc.accumulator().add(state, new Object[]{arg});
+                }
+            }
+
+            acc.accumulator().end(state, result);
+
+            output[i + cardinality] = acc.convertResult(result.get());
```

Comment:
Can `allFields` be replaced with mapping?
`i + cardinality` can be `mapping.get(i)`

Reply:
Mapping would make things even more confusing, because there are the following lines at the end of this method.
```
if (groupId != NO_GROUP_ID && type == AggregateType.MAP) {
   output[output.length - 1] = groupId;
}

```